[RenderStyleGen] Generate initial value functions

weinig opened the pull request:
#### c068d5d6bd1cf26b239439550969a4cbe92dbb64
<pre>
[RenderStyleGen] Generate initial value functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=303785">https://bugs.webkit.org/show_bug.cgi?id=303785</a>

Reviewed by Darin Adler.

Adds support for generating initial value functions for RenderStyle
based on the &apos;initial&apos; entries in CSSProperties.json.

This works via a very crude &quot;parser&quot; in process-css-properties.py
for the &apos;initial&apos; string, which handles just keywords and numeric
values. Most of the properties are now generated, but a few problems
remain that require custom definitions, including:
  - &lt;integer&gt; productions are a problem, as we don&apos;t know ahead of time if
    a raw number should be foo_css_number or foo_css_integer. Supporting
    foo_css_number for Style::Integer&lt;&gt; is one potential solution.
  - A few enum properties have enum values that don&apos;t exactly match the
    associated keyword.
  - The BorderImage and MaskImage sub-properties are not setup for literals
    at all.
  - There is not yet enough information to initialize EnumSet properties.
    Teaching EnumSet properties to handle CSS::Keyword::* types would
    solve this.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesInitialInlines.h: Added.
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStylePropertiesInitialCustom.cpp: Added.
* Source/WebCore/rendering/style/RenderStylePropertiesInitialCustom.h: Added.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBoxData.cpp:
* Source/WebCore/rendering/style/StyleBoxData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.cpp:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.h:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.cpp:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h:
* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/svg/StyleSVGStrokeDashoffset.h:

Canonical link: <a href="https://commits.webkit.org/304202@main">https://commits.webkit.org/304202@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/385f5a2780e56c6c939b32d6a19e2232d95c08fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134867 "19 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46152 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142378 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136736 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7140 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103053 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137813 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5586 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/120916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83894 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5415 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3030 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2971 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39072 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145076 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6968 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39647 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7026 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5846 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111779 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28361 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5253 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117199 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60897 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7015 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6798 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70593 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7034 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6908 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
```diff
@@ -199,14 +199,14 @@ void BoxTreeUpdater::adjustStyleIfNeeded(const RenderElement& renderer, RenderSt
             // This looks like continuation renderer.
             if (shouldNotRetainBorderPaddingAndMarginStart) {
-                styleToAdjust.setMarginStart(RenderStyle::initialMargin());
+                styleToAdjust.setMarginStart(RenderStyle::initialMarginLeft());
```
Surprised that this isn’t `initialMarginStart`.

weinig replied:
There are no defined initial functions for the logical values. Fortunately, all the margin edges have the same initial value.

weinig replied:
But a comment is warranted.

darinadler commented on Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
```diff
@@ -199,14 +199,14 @@ void BoxTreeUpdater::adjustStyleIfNeeded(const RenderElement& renderer, RenderSt
             }
             if (shouldNotRetainBorderPaddingAndMarginEnd) {
-                styleToAdjust.setMarginEnd(RenderStyle::initialMargin());
+                styleToAdjust.setMarginEnd(RenderStyle::initialMarginRight());
```
Surprised that this isn’t `initialMarginEnd`.

darinadler commented on Source/WebCore/rendering/style/RenderStylePropertiesInitialCustom.h:
```diff
@@ -0,0 +1,247 @@
+
+namespace WebCore {
+
+constexpr Style::MathDepth RenderStyleProperties::initialMathDepth()
```
Do we have to repeat the `constexpr` in these function definitions given the functions were already declared that way in the class definition?

darinadler commented on Source/WebCore/style/values/borders/StyleCornerShapeValue.h:
```diff
@@ -40,26 +40,27 @@ using SuperellipseFunction = FunctionNotation<CSSValueSuperellipse, Number<CSS::
+    constexpr CornerShapeValue(CSS::Keyword::Notch) : superellipse { 0.0 } { }
+    constexpr CornerShapeValue(CSS::Keyword::Straight) : superellipse { std::numeric_limits<double>::infinity() } { }
+    constexpr CornerShapeValue(CSS::Keyword::Squircle) : superellipse { 4.0 } { }
+    constexpr CornerShapeValue(SuperellipseFunction value) : superellipse { value } { }
```
Should this use `std::move` or `WTFMove`?

weinig replied:
Nah, Superellispe is just a thin wrapper around a double.

darinadler commented on Source/WebCore/style/values/svg/StyleSVGStrokeDashoffset.h:
```diff
@@ -43,6 +43,7 @@ struct SVGStrokeDashoffset {
     SVGStrokeDashoffsetLength value;
 
     SVGStrokeDashoffset(SVGStrokeDashoffsetLength&& length) : value { WTFMove(length) } { }
+    SVGStrokeDashoffset(CSS::ValueLiteral<CSS::NumberUnit::Number> literal) : value { Fixed { static_cast<float>(literal.value) } } { }
```
Cast to `float` is so ugly. No doubt it’s needed.

weinig replied:
Yeah, compiler can’t figure it out with out it.

weinig committed:
[RenderStyleGen] Generate initial value functions
https://bugs.webkit.org/show_bug.cgi?id=303785

Reviewed by Darin Adler.

Adds support for generating initial value functions for RenderStyle
based on the 'initial' entries in CSSProperties.json.

This works via a very crude "parser" in process-css-properties.py
for the 'initial' string, which handles just keywords and numeric
values. Most of the properties are now generated, but a few problems
remain that require custom definitions, including:
  - <integer> productions are a problem, as we don't know ahead of time if
    a raw number should be foo_css_number or foo_css_integer. Supporting
    foo_css_number for Style::Integer<> is one potential solution.
  - A few enum properties have enum values that don't exactly match the
    associated keyword.
  - The BorderImage and MaskImage sub-properties are not setup for literals
    at all.
  - There is not yet enough information to initialize EnumSet properties.
    Teaching EnumSet properties to handle CSS::Keyword::* types would
    solve this.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStylePropertiesInitialInlines.h: Added.
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/layout/integration/LayoutIntegrationBoxTreeUpdater.cpp:
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStylePropertiesInitialCustom.cpp: Added.
* Source/WebCore/rendering/style/RenderStylePropertiesInitialCustom.h: Added.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBoxData.cpp:
* Source/WebCore/rendering/style/StyleBoxData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.cpp:
* Source/WebCore/style/values/borders/StyleCornerShapeValue.h:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.cpp:
* Source/WebCore/style/values/color-adjust/StyleColorScheme.h:
* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/svg/StyleSVGStrokeDashoffset.h:

Canonical link: https://commits.webkit.org/304202@main
